File: lambda_cbn.cpp
#include <iomanip>
#include <iostream>
#include <optional>
#include <variant>
#include <cassert>
#include <string>
#include <memory>
#include <unordered_map>
#include <unordered_set>
#include "container.hpp"
#include "slice.hpp"
#ifndef USE_GMP
#include "strint.hpp"
#else
#include "gmpint.hpp"
#endif
#if defined _WIN32
#include <Windows.h>
#elif defined __unix__
#include <unistd.h>
#include <signal.h>
#include <sys/resource.h>
#endif
#ifndef STACK_SIZE
#define STACK_SIZE 8388608 // 8 MiB
#endif
char *stack_top;
char *stack_cur;
void ini_stack() {
    char dummy;
    stack_top = &dummy;
}
bool chk_stack() {
    char dummy;
    stack_cur = &dummy;
    return stack_top - stack_cur >= STACK_SIZE / 2;
}
#if defined _WIN32
void clr_flag() {
    GetAsyncKeyState(VK_ESCAPE);
}
bool chk_flag() {
    return GetAsyncKeyState(VK_ESCAPE) & 0x0001;
}
#elif defined __unix__
bool flag_rec;
void set_flag(int) {
    flag_rec = 1;
}
void clr_flag() {
    flag_rec = 0;
}
bool chk_flag() {
    return flag_rec;
}
#endif
auto read(Slice &exp) {
    auto i = exp.get_beg();
    auto n = exp.get_end();
    for (;; i++) {
        if (i == n) {
            return exp.reset_to(i, n), exp.from_to(i, i);
        } else if (*i != ' ') {
            break;
        }
    }
    auto j = i;
    auto c = 0;
    for (;; i++) {
        if ((i == n || *i == ' ') && c == 0) {
            return exp.reset_to(i, n), exp.from_to(j, i);
        } else if (i == n) {
            throw std::runtime_error("mismatched parentheses");
        } else if (*i == '(') {
            c++;
        } else if (*i == ')') {
            c--;
        }
    }
}
StrInt operator+(StrInt const &lval, StrInt const &rval);
StrInt operator-(StrInt const &lval, StrInt const &rval);
StrInt operator*(StrInt const &lval, StrInt const &rval);
StrInt operator/(StrInt const &lval, StrInt const &rval);
StrInt operator%(StrInt const &lval, StrInt const &rval);
bool operator>(StrInt const &lval, StrInt const &rval);
bool operator<(StrInt const &lval, StrInt const &rval);
bool operator>=(StrInt const &lval, StrInt const &rval);
bool operator<=(StrInt const &lval, StrInt const &rval);
bool operator==(StrInt const &lval, StrInt const &rval);
bool operator!=(StrInt const &lval, StrInt const &rval);
typedef StrInt (*opr_t)(StrInt const &, StrInt const &);
typedef bool (*cmp_t)(StrInt const &, StrInt const &);
static inline std::unordered_map<char, opr_t> const oprs = {
    {'+', operator+},
    {'-', operator-},
    {'*', operator*},
    {'/', operator/},
    {'%', operator%},
};
static inline std::unordered_map<char, cmp_t> const cmps = {
    {'>', operator>},
    {'<', operator<},
    {'=', operator==},
};
class Tree {
    enum TokenIdx: std::size_t {
        Ini, Par,
        Nil, Int,
        Opr, AOI,
        Cmp, ACI,
        LEF, EEF, // Lazy/Eager-Evaluation Function
        App, Arg,
    };
    using TokenVar = std::variant<
        std::nullopt_t, std::string,
        std::monostate, StrInt,
        std::pair<char, opr_t>, std::pair<std::pair<char, opr_t>, StrInt>,
        std::pair<char, cmp_t>, std::pair<std::pair<char, cmp_t>, StrInt>,
        Box<std::pair<std::string, Tree>>, Box<std::pair<std::string, Tree>>,
        Box<std::pair<Tree, Tree>>, std::shared_ptr<std::pair<Tree, bool>>>;
    TokenVar token;
    template <typename... Args, typename = std::enable_if_t<std::is_constructible_v<TokenVar, Args &&...>>>
    Tree(Args &&...args): token(std::forward<Args>(args)...) {}
    friend Box<std::pair<std::string, Tree>>;
    friend Box<std::pair<Tree, Tree>>;
    static Tree first(Tree &&fst) {
        if (fst.token.index() == TokenIdx::Ini) {
            throw std::runtime_error("empty expression");
        } else {
            return std::move(fst);
        }
    }
    static Tree build(Tree &&fst, Tree &&snd) {
        if (fst.token.index() == TokenIdx::Ini) {
            return std::move(snd);
        } else {
            return Box<std::pair<Tree, Tree>>::make(std::move(fst), std::move(snd));
        }
    }
    // static Tree parse(Slice &&exp, Tree &&fst = std::nullopt) {
    //     if (auto sym = read(exp); sym.empty()) {
    //         return first(std::move(fst));
    //     } else if (sym[0] == '\\') {
    //         return build(std::move(fst), Tree(std::in_place_index<TokenIdx::LEF>, Box<std::pair<std::string, Tree>>::make(sym(1, 0), parse(std::move(exp)))));
    //     } else if (sym[0] == '|') {
    //         return parse(std::move(exp), Tree(std::in_place_index<TokenIdx::LEF>, Box<std::pair<std::string, Tree>>::make(sym(1, 0), first(std::move(fst)))));
    //     } else if (sym[0] == '^') {
    //         return build(std::move(fst), Tree(std::in_place_index<TokenIdx::EEF>, Box<std::pair<std::string, Tree>>::make(sym(1, 0), parse(std::move(exp)))));
    //     } else if (sym[0] == '@') {
    //         return parse(std::move(exp), Tree(std::in_place_index<TokenIdx::EEF>, Box<std::pair<std::string, Tree>>::make(sym(1, 0), first(std::move(fst)))));
    //     } else {
    //         return parse(std::move(exp), build(std::move(fst), lex(std::move(sym))));
    //     }
    // }
    static Tree parse(Slice &&exp, Tree &&fun = std::nullopt, Tree &&fst = std::nullopt) {
        if (auto sym = read(exp); sym.empty()) {
            return build(std::move(fun), first(std::move(fst)));
        } else if (sym[0] == '\\') {
            return build(std::move(fun), build(std::move(fst), Tree(std::in_place_index<TokenIdx::LEF>, Box<std::pair<std::string, Tree>>::make(sym(1, 0), parse(std::move(exp))))));
        } else if (sym[0] == '|') {
            return parse(std::move(exp), Tree(std::in_place_index<TokenIdx::LEF>, Box<std::pair<std::string, Tree>>::make(sym(1, 0), build(std::move(fun), first(std::move(fst))))));
        } else if (sym[0] == '^') {
            return build(std::move(fun), build(std::move(fst), Tree(std::in_place_index<TokenIdx::EEF>, Box<std::pair<std::string, Tree>>::make(sym(1, 0), parse(std::move(exp))))));
        } else if (sym[0] == '@') {
            return parse(std::move(exp), Tree(std::in_place_index<TokenIdx::EEF>, Box<std::pair<std::string, Tree>>::make(sym(1, 0), build(std::move(fun), first(std::move(fst))))));
        } else {
            return parse(std::move(exp), std::move(fun), build(std::move(fst), lex(std::move(sym))));
        }
    }
    static Tree lex(Slice const &sym) {
        if (sym[0] == '(' && sym[-1] == ')') {
            return parse(sym(1, -1));
        } else if (sym[0] == '$') {
            return Tree(std::in_place_index<TokenIdx::Par>, sym(1, 0));
        } else if (sym.size() == 3 && sym == "...") {
            return Tree(std::in_place_index<TokenIdx::Nil>);
        } else if (auto const &o = oprs.find(sym[0]); sym.size() == 1 && o != oprs.end()) {
            return Tree(std::in_place_index<TokenIdx::Opr>, *o);
        } else if (auto const &c = cmps.find(sym[0]); sym.size() == 1 && c != cmps.end()) {
            return Tree(std::in_place_index<TokenIdx::Cmp>, *c);
        } else try {
            return Tree(std::in_place_index<TokenIdx::Int>, StrInt::from_string(sym));
        } catch (...) {
            throw std::runtime_error("invalid symbol: " + std::string(sym));
        }
    }
    void calc() {
        if (chk_stack()) {
            throw std::runtime_error("recursion limit exceeded");
        }
    tail_call:
        if (chk_flag()) {
            throw std::runtime_error("keyboard interrupt");
        }
        if (auto papp = std::get_if<TokenIdx::App>(&token)) {
            auto &[fst, snd] = **papp;
            fst.calc();
            if (auto pnil = std::get_if<TokenIdx::Nil>(&fst.token)) {
                token.emplace<TokenIdx::Nil>();
            } else if (auto plef = std::get_if<TokenIdx::LEF>(&fst.token)) {
                auto &[par, tmp] = **plef;
                tmp.substitute(std::make_shared<std::pair<Tree, bool>>(std::move(snd), 0), std::move(par));
                *this = Tree(std::move(tmp));
                goto tail_call;
            } else if (auto peef = std::get_if<TokenIdx::EEF>(&fst.token)) {
                snd.calc();
                auto &[par, tmp] = **peef;
                tmp.substitute(std::make_shared<std::pair<Tree, bool>>(std::move(snd), 1), std::move(par));
                *this = Tree(std::move(tmp));
                goto tail_call;
            } else if (auto popr = std::get_if<TokenIdx::Opr>(&fst.token)) {
                snd.calc();
                if (auto pint = std::get_if<TokenIdx::Int>(&snd.token); pint && (*pint || popr->first != '/' && popr->first != '%')) {
                    token = std::make_pair(std::move(*popr), std::move(*pint));
                } else {
                    throw std::runtime_error("cannot apply " + fst.translate() + " on: " + snd.translate());
                }
            } else if (auto pcmp = std::get_if<TokenIdx::Cmp>(&fst.token)) {
                snd.calc();
                if (auto pint = std::get_if<TokenIdx::Int>(&snd.token)) {
                    token = std::make_pair(std::move(*pcmp), std::move(*pint));
                } else {
                    throw std::runtime_error("cannot apply " + fst.translate() + " on: " + snd.translate());
                }
            } else if (auto paoi = std::get_if<TokenIdx::AOI>(&fst.token)) {
                snd.calc();
                if (auto pint = std::get_if<TokenIdx::Int>(&snd.token)) {
                    token = paoi->first.second(std::move(*pint), std::move(paoi->second));
                } else {
                    throw std::runtime_error("cannot apply " + fst.translate() + " on: " + snd.translate());
                }
            } else if (auto paci = std::get_if<TokenIdx::ACI>(&fst.token)) {
                snd.calc();
                if (auto pint = std::get_if<TokenIdx::Int>(&snd.token)) {
                    static const auto T = Tree(std::in_place_index<TokenIdx::LEF>, Box<std::pair<std::string, Tree>>::make("T", Tree(std::in_place_index<TokenIdx::LEF>, Box<std::pair<std::string, Tree>>::make("F", Tree(std::in_place_index<TokenIdx::Par>, "T")))));
                    static const auto F = Tree(std::in_place_index<TokenIdx::LEF>, Box<std::pair<std::string, Tree>>::make("T", Tree(std::in_place_index<TokenIdx::LEF>, Box<std::pair<std::string, Tree>>::make("F", Tree(std::in_place_index<TokenIdx::Par>, "F")))));
                    *this = paci->first.second(std::move(*pint), std::move(paci->second)) ? T : F;
                } else {
                    throw std::runtime_error("cannot apply " + fst.translate() + " on: " + snd.translate());
                }
            } else {
                throw std::runtime_error("invalid function: " + fst.translate());
            }
        } else if (auto parg = std::get_if<TokenIdx::Arg>(&token)) {
            auto &shr = (*parg)->first;
            auto rec = (*parg)->second;
            if (parg->use_count() == 1) {
                *this = Tree(std::move(shr));
                if (not rec) {
                    goto tail_call;
                }
            } else {
                if (not rec) {
                    shr.calc();
                    rec = true;
                }
                *this = shr;
            }
        }
    }
    void substitute(std::shared_ptr<std::pair<Tree, bool>> const &arg, std::string const &tar) {
        if (auto papp = std::get_if<TokenIdx::App>(&token)) {
            auto &[fst, snd] = **papp;
            fst.substitute(arg, tar);
            snd.substitute(arg, tar);
        } else if (auto plef = std::get_if<TokenIdx::LEF>(&token)) {
            auto &[par, tmp] = **plef;
            if (par != tar) {
                tmp.substitute(arg, tar);
            }
        } else if (auto peef = std::get_if<TokenIdx::EEF>(&token)) {
            auto &[par, tmp] = **peef;
            if (par != tar) {
                tmp.substitute(arg, tar);
            }
        } else if (auto ppar = std::get_if<TokenIdx::Par>(&token)) {
            if (*ppar == tar) {
                token.emplace<TokenIdx::Arg>(arg);
            }
        }
    }
    void substitute(std::unordered_map<std::string, std::shared_ptr<std::pair<Tree, bool>>> &map) {
        if (auto papp = std::get_if<TokenIdx::App>(&token)) {
            auto &[fst, snd] = **papp;
            fst.substitute(map);
            snd.substitute(map);
        } else if (auto plef = std::get_if<TokenIdx::LEF>(&token)) {
            auto &[par, tmp] = **plef;
            if (auto const &it = map.find(par); it != map.end()) {
                auto sav = std::move(it->second);
                tmp.substitute(map);
                it->second = std::move(sav);
            } else {
                tmp.substitute(map);
            }
        } else if (auto peef = std::get_if<TokenIdx::EEF>(&token)) {
            auto &[par, tmp] = **peef;
            if (auto const &it = map.find(par); it != map.end()) {
                auto sav = std::move(it->second);
                tmp.substitute(map);
                it->second = std::move(sav);
            } else {
                tmp.substitute(map);
            }
        } else if (auto ppar = std::get_if<TokenIdx::Par>(&token)) {
            if (auto const &it = map.find(*ppar); it != map.end() && it->second) {
                token.emplace<TokenIdx::Arg>(it->second);
            }
        }
    }
    void analyze(std::unordered_set<std::string> &set) const {
        if (auto papp = std::get_if<TokenIdx::App>(&token)) {
            auto &[fst, snd] = **papp;
            fst.analyze(set);
            snd.analyze(set);
        } else if (auto plef = std::get_if<TokenIdx::LEF>(&token)) {
            auto &[par, tmp] = **plef;
            if (auto const &it = set.find(par); it != set.end()) {
                tmp.analyze(set);
            } else {
                auto jt = set.insert(par).first;
                tmp.analyze(set);
                set.erase(jt);
            }
        } else if (auto peef = std::get_if<TokenIdx::EEF>(&token)) {
            auto &[par, tmp] = **peef;
            if (auto const &it = set.find(par); it != set.end()) {
                tmp.analyze(set);
            } else {
                auto jt = set.insert(par).first;
                tmp.analyze(set);
                set.erase(jt);
            }
        } else if (auto ppar = std::get_if<TokenIdx::Par>(&token)) {
            if (auto const &it = set.find(*ppar); it == set.end()) {
                throw std::runtime_error("unbound variable: $" + *ppar);
            }
        }
    }
    static inline std::unordered_map<std::string, std::shared_ptr<std::pair<Tree, bool>>> map;
public:
    static auto const &put(Slice &&exp, std::string const &par, bool calc) {
        auto res = parse(std::move(exp));
        res.substitute(map);
        std::unordered_set<std::string> set;
        res.analyze(set);
        map.erase(par);
        if (calc) {
            clr_flag();
            res.calc();
        }
        return map.emplace(par, std::make_shared<std::pair<Tree, bool>>(std::move(res), calc)).first->second->first;
    }
    static auto const &dir() {
        return map;
    }
    static void clr() {
        return map.clear();
    }
    std::string translate(bool lb = 0, bool rb = 0) const {
        if (auto pnil = std::get_if<TokenIdx::Nil>(&token)) {
            return "...";
        } else if (auto plef = std::get_if<TokenIdx::LEF>(&token)) {
            auto &[par, tmp] = **plef;
            auto s = "\\" + par + " " + tmp.translate(0, rb && !rb);
            return rb ? "(" + s + ")" : s;
        } else if (auto peef = std::get_if<TokenIdx::EEF>(&token)) {
            auto &[par, tmp] = **peef;
            auto s = "^" + par + " " + tmp.translate(0, rb && !rb);
            return rb ? "(" + s + ")" : s;
        } else if (auto pint = std::get_if<TokenIdx::Int>(&token)) {
            return pint->to_string();
        } else if (auto popr = std::get_if<TokenIdx::Opr>(&token)) {
            return std::string{popr->first};
        } else if (auto pcmp = std::get_if<TokenIdx::Cmp>(&token)) {
            return std::string{pcmp->first};
        } else if (auto paoi = std::get_if<TokenIdx::AOI>(&token)) {
            auto s = std::string{paoi->first.first, ' '} + paoi->second.to_string();
            return lb ? "(" + s + ")" : s;
        } else if (auto paci = std::get_if<TokenIdx::ACI>(&token)) {
            auto s = std::string{paci->first.first, ' '} + paci->second.to_string();
            return lb ? "(" + s + ")" : s;
        } else if (auto papp = std::get_if<TokenIdx::App>(&token)) {
            auto &[fst, snd] = **papp;
            auto s = fst.translate(lb && !lb, 1) + " " + snd.translate(1, rb && !lb);
            return lb ? "(" + s + ")" : s;
        } else if (auto parg = std::get_if<TokenIdx::Arg>(&token)) {
            return (*parg)->first.translate(lb, rb);
        } else if (auto ppar = std::get_if<TokenIdx::Par>(&token)) {
            return "$" + *ppar;
        } else {
            assert(false); // unreachable
        }
    }
};
int main(int argc, char *argv[]) {
    ini_stack();
    bool check_stdin = false;
    bool check_stdout = false;
    bool check_stderr = false;
#if defined _WIN32
    DWORD dwModeTemp;
    check_stdin = GetConsoleMode(GetStdHandle(STD_INPUT_HANDLE), &dwModeTemp);
    check_stdout = GetConsoleMode(GetStdHandle(STD_OUTPUT_HANDLE), &dwModeTemp);
    check_stderr = GetConsoleMode(GetStdHandle(STD_ERROR_HANDLE), &dwModeTemp);
#elif defined __unix__
    check_stdin = isatty(fileno(stdin));
    check_stdout = isatty(fileno(stdout));
    check_stderr = isatty(fileno(stderr));
    struct rlimit rlim;
    getrlimit(RLIMIT_STACK, &rlim);
    rlim.rlim_cur = STACK_SIZE;
    setrlimit(RLIMIT_STACK, &rlim);
    struct sigaction act;
    act.sa_handler = set_flag;
    sigemptyset(&act.sa_mask);
    act.sa_flags = 0; // use SA_RESTART to avoid getting EOF when SIGINT is received during input
    sigaction(SIGINT, &act, NULL);
#endif
    std::string ps_in = check_stderr && check_stdin ? ">> " : "";
    std::string ps_out = check_stderr && check_stdout ? "=> " : "";
    std::string ps_res = check_stderr && check_stdout ? "== " : "";
    for (bool end = false; not end;) {
        std::cerr << ps_in;
        Slice exp = Slice::getline(std::cin);
        if (std::cin.eof()) {
            end = true;
            if (check_stderr && check_stdin) {
                std::cerr << std::endl;
            }
        }
        try {
            if (auto cmd = read(exp); cmd.empty() || cmd.size() == 1 && cmd == "#") {
                continue;
            } else if (cmd[0] == ':') {
                Tree::put(std::move(exp), cmd(1, 0), 0);
            } else if (cmd.size() == 3 && cmd == "cal") {
                auto &res = Tree::put(std::move(exp), "", 1);
                std::cerr << ps_res;
                std::cout << res.translate() << std::endl;
            } else if (cmd.size() == 3 && cmd == "dir") {
                for (auto &[par, arg] : Tree::dir()) {
                    std::cerr << ps_out;
                    std::cout << ":" + par << std::endl;
                }
            } else if (cmd.size() == 3 && cmd == "clr") {
                Tree::clr();
            } else if (cmd.size() == 3 && cmd == "end") {
                end = true;
            } else {
                throw std::runtime_error("unknown command: " + std::string(cmd));
            }
        } catch (std::exception const &e) {
            std::cerr << "Runtime Error: " << e.what() << std::endl;
        }
    }
    return 0;
}

File: lambda_cbn.rs
#![feature(box_patterns)]
#![feature(if_let_guard)]
use std::arch::asm;
use std::cell::RefCell;
use std::collections::HashMap;
use std::io;
use std::mem;
use std::rc::Rc;
extern crate atty;
extern crate lazy_static;
extern crate num_bigint;
extern crate num_traits;
use atty::Stream;
use lazy_static::lazy_static;
use num_bigint::BigInt;
use num_traits::Zero;
lazy_static! {
    static ref PS_IN: &'static str = if atty::is(Stream::Stderr) && atty::is(Stream::Stdin) { ">> " } else { "" };
    static ref PS_OUT: &'static str = if atty::is(Stream::Stderr) && atty::is(Stream::Stdout) { "=> " } else { "" };
    static ref PS_RES: &'static str = if atty::is(Stream::Stderr) && atty::is(Stream::Stdout) { "== " } else { "" };
    static ref ADD_LN: bool = if atty::is(Stream::Stderr) && atty::is(Stream::Stdin) { true } else { false };
    static ref STACK_MAX: usize = 0o4000_0000;
    static ref STACK_TOP: usize = stack_ptr();
}
fn stack_ptr() -> usize {
    unsafe {
        let mut ptr: usize;
        asm!("mov {}, rsp", out(reg) ptr);
        ptr
    }
}
fn stack_err() -> bool {
    *STACK_TOP - *STACK_MAX / 2 > stack_ptr()
}
fn read(exp: &str) -> Result<(&str, &str), String> {
    let i = exp.as_bytes().iter().position(|&c| c != b' ').unwrap_or(exp.len());
    let mut ctr: usize = 0;
    for (j, &c) in exp.as_bytes().iter().enumerate().skip(i) {
        if ctr == 0 && c == b' ' {
            return Ok((&exp[i..j], &exp[j..]));
        } else if c == b'(' {
            ctr += 1;
        } else if c == b')' {
            ctr -= 1;
        }
    }
    if ctr != 0 {
        Err("unmatched parenthesis".to_string())
    } else {
        Ok((&exp[i..], ""))
    }
}
type Oprt = fn(BigInt, BigInt) -> BigInt;
type Cmpt = fn(BigInt, BigInt) -> bool;
static OPRS: [(&str, Oprt); 5] = [
    ("+", |a, b| a + b),
    ("-", |a, b| a - b),
    ("*", |a, b| a * b),
    ("/", |a, b| a / b),
    ("%", |a, b| a % b),
];
static CMPS: [(&str, Cmpt); 3] = [
    (">", |a, b| a > b),
    ("<", |a, b| a < b),
    ("=", |a, b| a == b),
];
#[derive(Clone)]
enum Tree {
    Empty,
    Ellipsis,
    Int(BigInt),
    Opr(Oprt, String),
    Cmp(Cmpt, String),
    OprInt(Oprt, String, BigInt),
    CmpInt(Cmpt, String, BigInt),
    Fun(String, Box<Tree>),
    Out(String, Box<Tree>),
    App(Box<Tree>, Box<Tree>),
    Arg(Rc<RefCell<(Tree, bool)>>),
    Par(String),
}
impl Tree {
    fn first(fst: Tree) -> Result<Tree, String> {
        match fst {
            Tree::Empty => Err("empty expression".to_string()),
            _ => Ok(fst),
        }
    }
    fn build(fst: Tree, snd: Tree) -> Result<Tree, String> {
        match fst {
            Tree::Empty => Ok(snd),
            _ => Ok(Tree::App(Box::new(fst), Box::new(snd))),
        }
    }
    fn parse(exp: &str, fun: Tree, fst: Tree) -> Result<Tree, String> {
        let (sym, rest) = read(exp)?;
        if sym.is_empty() {
            Self::build(fun, Self::first(fst)?)
        } else if sym.starts_with('\\') {
            let tmp = Tree::Fun(sym[1..].to_string(), Box::new(Self::parse(rest, Tree::Empty, Tree::Empty)?));
            Self::build(fun, Self::build(fst, tmp)?)
        } else if sym.starts_with('|') {
            let tmp = Tree::Fun(sym[1..].to_string(), Box::new(Self::build(fun, Self::first(fst)?)?));
            Self::parse(rest, tmp, Tree::Empty)
        } else if sym.starts_with('^') {
            let tmp = Tree::Out(sym[1..].to_string(), Box::new(Self::parse(rest, Tree::Empty, Tree::Empty)?));
            Self::build(fun, Self::build(fst, tmp)?)
        } else if sym.starts_with('@') {
            let tmp = Tree::Out(sym[1..].to_string(), Box::new(Self::build(fun, Self::first(fst)?)?));
            Self::parse(rest, tmp, Tree::Empty)
        } else {
            Self::parse(rest, fun, Self::build(fst, Self::lex(sym)?)?)
        }
    }
    fn lex(sym: &str) -> Result<Tree, String> {
        if sym.starts_with('(') && sym.ends_with(')') {
            Self::parse(&sym[1..sym.len() - 1], Tree::Empty, Tree::Empty)
        } else if sym.starts_with('$') {
            Ok(Tree::Par(sym[1..].to_string()))
        } else if sym == "..." {
            Ok(Tree::Ellipsis)
        } else if let Some((name, opr)) = OPRS.iter().find(|(s, _)| *s == sym) {
            Ok(Tree::Opr(*opr, name.to_string()))
        } else if let Some((name, cmp)) = CMPS.iter().find(|(s, _)| *s == sym) {
            Ok(Tree::Cmp(*cmp, name.to_string()))
        } else if let Ok(num) = sym.parse::<BigInt>() {
            Ok(Tree::Int(num))
        } else {
            Err(format!("unknown symbol: {}", sym))
        }
    }
    fn calculate(self) -> Result<Self, String> {
        if stack_err() {
            return Err("recursion limit exceeded".to_string());
        }
        match self {
            Tree::App(box fst, box snd) => match fst.calculate()? {
                Tree::Ellipsis => Ok(Tree::Ellipsis),
                Tree::Fun(sym, box mut tmp) => {
                    tmp.substitute(&Rc::new(RefCell::new((snd, false))), &sym);
                    tmp.calculate()
                }
                Tree::Out(sym, box mut tmp) => {
                    let snd = snd.calculate()?;
                    tmp.substitute(&Rc::new(RefCell::new((snd, true))), &sym);
                    tmp.calculate()
                }
                Tree::Opr(opr, name) => match snd.calculate()? {
                    Tree::Int(int) if !int.is_zero() || (name != "/" && name != "%") => Ok(Tree::OprInt(opr, name, int)),
                    snd => Err(format!("cannot apply {} on: {}", name, snd.translate(false, false))),
                },
                Tree::Cmp(cmp, name) => match snd.calculate()? {
                    Tree::Int(int) => Ok(Tree::CmpInt(cmp, name, int)),
                    snd => Err(format!("cannot apply {} on: {}", name, snd.translate(false, false))),
                },
                Tree::OprInt(opr, name, int) => match snd.calculate()? {
                    Tree::Int(val) => Ok(Tree::Int(opr(val, int))),
                    snd => Err(format!("cannot apply {} {} on: {}", name, int, snd.translate(false, false))),
                },
                Tree::CmpInt(cmp, name, int) => match snd.calculate()? {
                    Tree::Int(val) => Ok(Tree::Fun("T".to_string(), Box::new(Tree::Fun("F".to_string(), Box::new(Tree::Par(if cmp(val, int) { "T" } else { "F" }.to_string())))))),
                    snd => Err(format!("cannot apply {} {} on: {}", name, int, snd.translate(false, false))),
                },
                fst => Err(format!("invalid function: {}", fst.translate(false, false))),
            },
            Tree::Arg(arg) => {
                let (shr, rec) = &mut *arg.borrow_mut();
                if !*rec {
                    *shr = mem::replace(shr, Tree::Empty).calculate()?;
                    *rec = true;
                }
                Ok(if Rc::strong_count(&arg) == 1 { mem::replace(shr, Tree::Empty) } else { shr.clone() })
            }
            Tree::Par(key) => {
                Err(format!("unbound variable: ${}", key))
            }
            _ => Ok(self),
        }
    }
    fn substitute(&mut self, arg: &Rc<RefCell<(Tree, bool)>>, par: &String) {
        match self {
            Tree::Fun(sym, box tmp) if sym != par => {
                tmp.substitute(arg, par);
            }
            Tree::Out(sym, box tmp) if sym != par => {
                tmp.substitute(arg, par);
            }
            Tree::App(box fst, box snd) => {
                fst.substitute(arg, par);
                snd.substitute(arg, par);
            }
            Tree::Par(sym) if sym == par => {
                *self = Tree::Arg(arg.clone());
            }
            _ => {}
        }
    }
    fn translate(&self, lb: bool, rb: bool) -> String {
        match self {
            Tree::Fun(sym, box tree) => {
                let xxx = format!("\\{} {}", sym, tree.translate(false, rb && !rb));
                if rb { format!("({})", xxx) } else { xxx }
            }
            Tree::Out(sym, box tree) => {
                let xxx = format!("^{} {}", sym, tree.translate(false, rb && !rb));
                if rb { format!("({})", xxx) } else { xxx }
            }
            Tree::Ellipsis => "...".to_string(),
            Tree::Int(i) => i.to_string(),
            Tree::Opr(_, name) => name.clone(),
            Tree::Cmp(_, name) => name.clone(),
            Tree::OprInt(_, name, int) => {
                let xxx = format!("{} {}", name, int);
                if lb { format!("({})", xxx) } else { xxx }
            }
            Tree::CmpInt(_, name, int) => {
                let xxx = format!("{} {}", name, int);
                if lb { format!("({})", xxx) } else { xxx }
            }
            Tree::App(box fst, box snd) => {
                let xxx = format!("{} {}", fst.translate(lb && !lb, true), snd.translate(true, rb && !lb));
                if lb { format!("({})", xxx) } else { xxx }
            }
            Tree::Arg(arg) => {
                let (shr, rec) = &*arg.borrow();
                shr.translate(lb, rb)
            }
            Tree::Par(key) => format!("${}", key),
            _ => panic!("invalid tree"),
        }
    }
}
fn process(input: String, map: &mut HashMap<String, Rc<RefCell<(Tree, bool)>>>) -> Result<(), String> {
    let (cmd, exp) = read(&input)?;
    if cmd.is_empty() || cmd == "#" {
        Ok(())
    } else if cmd.starts_with(':') {
        let mut res = Tree::parse(exp, Tree::Empty, Tree::Empty)?;
        for (key, val) in map.iter() {
            res.substitute(val, key);
        }
        map.insert(cmd[1..].to_string(), Rc::new(RefCell::new((res, false))));
        Ok(())
    } else if cmd == "cal" {
        let mut res = Tree::parse(exp, Tree::Empty, Tree::Empty)?;
        for (key, val) in map.iter() {
            res.substitute(val, key);
        }
        let res = res.calculate()?;
        eprint!("{}", *PS_RES);
        println!("{}", res.translate(false, false));
        map.insert("".to_string(), Rc::new(RefCell::new((res, true))));
        Ok(())
    } else if cmd == "dir" {
        for (key, val) in map.iter() {
            eprint!("{}", *PS_OUT);
            println!(":{}", key);
        }
        Ok(())
    } else if cmd == "clr" {
        map.clear();
        Ok(())
    } else {
        Err(format!("unknown command: {}", cmd))
    }
}
fn main() {
    let mut map = HashMap::new();
    let mut nxt = true;
    while nxt {
        eprint!("{}", *PS_IN);
        let mut input = String::new();
        io::stdin().read_line(&mut input).unwrap();
        if input.ends_with('\r') {
            input.pop();
        }
        if input.ends_with('\n') {
            input.pop();
        } else if *ADD_LN {
            eprintln!();
            nxt = false;
        } else {
            nxt = false;
        }
        if input.ends_with('\r') {
            input.pop();
        }
        if let Err(err) = process(input, &mut map) {
            eprintln!("Runtime Error: {}", err);
        }
    }
}
